karapace/kafka_rest_apis/__init__.py

```diff
@@ -51,16 +51,48 @@ def __init__(self, config: Config) -> None:
         self._add_kafka_rest_routes()
         self.serializer = SchemaRegistrySerializer(config=config)
         self.proxies: Dict[str, "UserRestProxy"] = {}
+        self._proxy_lock = asyncio.Lock()
         log.info("REST proxy starting with (delegated authorization=%s)", self.config.get("rest_authorization", False))
+        self._idle_proxy_janitor_task: Optional[asyncio.Task] = None
 
     async def close(self) -> None:
+        if self._idle_proxy_janitor_task is not None:
+            self._idle_proxy_janitor_task.cancel()
+            self._idle_proxy_janitor_task = None
         async with AsyncExitStack() as stack:
             stack.push_async_callback(super().close)
             stack.push_async_callback(self.serializer.close)
 
             for proxy in self.proxies.values():
                 stack.push_async_callback(proxy.aclose)
 
+    async def _idle_proxy_janitor(self) -> None:
+        IDLE_PROXY_TIMEOUT = 5 * 60
+        while True:
+            await asyncio.sleep(IDLE_PROXY_TIMEOUT / 2)
+
+            key, proxy = None, None
+            async with self._proxy_lock:
+                # Always clean one at time per iteration, don't mutate dict while iterating
+                for _key, _proxy in self.proxies.items():
+                    # If UserRestProxy has consumers with state, disconnecting loses state
+                    if _proxy.num_consumers() > 0:
+                        timeout = self.config.get("consumer_idle_disconnect_timeout", 0)
+                        if timeout > 0 and _proxy.last_used + timeout < time.monotonic():
+                            key, proxy = _key, _proxy
+                            log.warning("Disconnecting idle consumers for %s", _proxy)
+                            break
+                    # If there are no consumers, connection can be recreated without losing any state
+                    else:
+                        if _proxy.last_used + IDLE_PROXY_TIMEOUT < time.monotonic():
```

Comment:
Not sure, but shouldn't this follow also the configuration option of disconnect time? If set to zero / cleaning disabled keep the proxy in the dictionary?

Reply:
I went without configuration option for this because without consumers the proxy is stateless, ie. it can be recreated, and the API user does not see any difference.